Add script build_datasets.py

Bankso opened the pull request:
Adds a script that creates and/or assigns files to Synapse Dataset entities based on entries in a MC2 Center Dataset Sharing Plan.
- takes a DSP CSV path or table SynId as input
- For each row in the DSP:
  - checks if a Dataset entity Synapse Id was provided and creates one, if needed
  - checks if files of a specific format should be added from the Synapse Id in column "Dataset Alias" or if the whole folder should be added
  - filters files, based on their format, if needed
  - adds filtered files or all folder contents to Dataset, depending on if filtering was performed
  - if a new Dataset entity was created, the new Dataset Synapse Id gets recorded as "DatasetView Key"
- Generates a new DSP CSV, if entries were added to "DatasetView Key" for any of the rows in the DSP

Bankso committed:
Create build_datasets.py

Bankso committed:
Update build_datasets.py

Add inline comments

vpchung commented on utils/build_datasets.py:
```diff
@@ -0,0 +1,167 @@
+
+    all_files = []
+    walk_path = synapseutils.walk(syn, scope, ["file"])
+    for dirpath, dirname, filename in walk_path:
```
nit: since `dirname` is not used, you can use `_` instead which will indicate it's a throwaway variable, e.g.

```python
    for dirpath, _, filename in walk_path:
```

Bankso replied:
Technically `dirpath` and `dirname` are throwaways, even though the code says the `dirpath` variable is used. I wasn't sure how to properly denote two throwaways in a single loop (I considered using `_` and `__` but that seemed strange) so I kept the names for both. 

What convention do you use to indicate more than one throwaway variable?

vpchung replied:
Ah, in that case, I would use this neat unpacking trick with the `*` operator!

```python
    for *_, filename in walk_path:
```

This essentially means that you are only interested in the last item of `walk_path`, discarding all the preceding items.

vpchung replied:
In case you're interested in more info: what the `*` operator does during unpacking is that it'll collect items into a list based on how the other items are collected.

Here's some different ways of how it could be used (using our names 😉):

```python
>>> persons = ["ashley", "orion", "aditi", "aditya", "verena"]
>>> pm, *team = persons
>>> pm
'ashley'
>>> team
['orion', 'aditi', 'aditya', 'verena']
>>>
>>> *mc2_team, cnb_team = persons
>>> mc2_team
['ashley', 'orion', 'aditi', 'aditya']
>>> cnb_team
'verena'
>>>
>>> pm, *mc2_devs, cnb_devs = persons
>>> pm
'ashley'
>>> mc2_devs
['orion', 'aditi', 'aditya']
>>> cnb_devs
'verena'
```

Bankso replied:
Awesome, thank you for the example, this makes sense!

vpchung commented on utils/build_datasets.py:
```diff
@@ -0,0 +1,167 @@
+    dataset_items = [
+        {"entityId": f, "versionNumber": syn.get(f, downloadFile=False).versionLabel}
+        for f in filtered_files
+    ]
```
Yessssss, I see you using comprehensions more now 😎 Very nice!

For this part, you could also build `dataset_items` directly during the walk rather than rely on the two temporary lists, e.g.

```python
    dataset_items = []
    walk_path = synapseutils.walk(syn, scope, ["file"])
    for dirpath, _, filename in walk_path:
        for f, entity_id in filename:
            if any(f.endswith(fmt) for fmt in formats):
                dataset_items.append({
                    "entityId": entity_id,
                    "versionNumber": syn.get(f, downloadFile=False).versionLabel
                })
```

where `any(...)` is a quick way to check if at least one item in the provided iterable is `True`, in this case, does the filename end with any of the listed formats?

Bankso replied:
This is great, thank you!

vpchung commented on utils/build_datasets.py:
```diff
@@ -0,0 +1,167 @@
+            dataset_name = row["DSP Dataset Name"]
+            formats = re.split(", |,", row["DSP Dataset File Formats"])
+            level = row["DSP Dataset Level"]
+            if level not in ["Metadata", "Auxiliary", "Not Applicable"]:
```
Alternatively, you can use do something like:

```python
            if level in ["Metadata", "Auxiliary", "Not Applicable"]:
                continue

            print(f"\nProcessing Dataset {dataset_name}")
            if len(dataset_id) > 0:
                ...
```

Where Python would just move on (`continue`) to the next iteration without doing the rest of the loop's code.   This would help reduce the number of indentation levels, since now the rest of the code can be un-indented by one level.

P.S. My personal limit on indentations is 3-4 levels -- anything beyond that, I'll start thinking that my code probably needs to split further into more functions.  But that's just me!  Go with your own preferred styling :-)

vpchung commented on utils/build_datasets.py:
```diff
@@ -0,0 +1,167 @@
+            level = row["DSP Dataset Level"]
+            if level not in ["Metadata", "Auxiliary", "Not Applicable"]:
+                print(f"\nProcessing Dataset {dataset_name}")
+                if len(dataset_id) > 0:  # check if a Dataset entity was previously recorded 
```
For this, you can directly evaluate the "truthy-ness" of `dataset_id` with:

```python
                if dataset_id:
```

This will return `True` so long as `dataset_id` is not falsy, that is: `0`, an empty string, `False`, or `None`

vpchung commented on utils/build_datasets.py:
```diff
@@ -0,0 +1,167 @@
+                    update_dsp_sheet = True  # record the new DatasetView_id in DSP
+                    print(f"--> New Dataset created!")
+
+                if len(formats) > 0:  # only filter files if formats were specified
```
Same feedback as above; in this case, it will return `True` so long as `formats` is not an empty list (along with the other falsy values).

vpchung reviewed (approved):
LGTM 🔥 

As always, I do have some suggestions, but no need to apply them.  I also didn't thoroughly test with an actual Synapse table unless you have a minimal example I could use to test with?  But from what I can tell, the code achieves what is needed.

Bankso committed:
Update build_datasets.py

Adjust filter_files_in_folder function to directly add filtered files to output list

Bankso committed:
Update build_datasets.py

Use continue statement to reduce nesting in main function and restructure indentations

Bankso committed:
Update build_datasets.py

Use *_ to discard unneeded variables

Bankso committed:
Update build_datasets.py

Remove len calls and directly evaluate if dataset_id and formats exist

Bankso committed:
Update build_datasets.py

Pass entity Syn Id to syn.get call and add message for when DSP entries are skipped

Bankso committed:
Update build_datasets.py

Re-enable function to store dataset
